### ISSUE TITLE
Interested in a Prometheus middleware?

### ISSUE THREAD

**Issue by cweagans** [go-macaron/macaron#161]
I've written a middleware for my Macaron application that exposes request counts/latency information to Prometheus. I'd be willing to open it up to the community if there's interest!

**Comment by Unknwon** [comment]
Yes, I'm personally interested in collecting such data in my application. 

My questions are:
- How do the APIs or use cases look like? 
- How easy it is to work with exiting metrics already exposed to Prometheus?

Thanks!

**Comment by cweagans** [comment]
Apologies for the non-response here. It kind of fell off my radar. Here is what we have:

```go
package handlers

import (
	"log"
	"reflect"
	"strconv"
	"time"

	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promauto"
	macaron "gopkg.in/macaron.v1"
)

var (
	// prometheusNamespace is used to prevent collision with other apis which use too generic a name.
        // TODO: This should be configurable by end-users of this middleware.
	prometheusNamespace = "my_application"

	responseLabels = []string{
		"method",
		"path",
	}
	responses = promauto.NewHistogramVec(
		prometheus.HistogramOpts{
			Name:      "response_duration_seconds",
			Help:      "The time spent on serving a response to the client",
			Namespace: prometheusNamespace,
			Buckets:   []float64{10, 100, 500, 1000, 5000, 30000, 60000},
		},
		responseLabels,
	)

	requestLabels = []string{
		"status",
		"method",
		"path",
	}
	requests = promauto.NewCounterVec(
		prometheus.CounterOpts{
			Name:      "request_total",
			Help:      "The number of requests received by this endpoint.",
			Namespace: prometheusNamespace,
		},
		requestLabels,
	)
)

// MetricsInvoker is an inject.FastInvoker wrapper so that we get a log.Logger.
type MetricsInvoker func(ctx *macaron.Context, log *log.Logger)

// Invoke is the entry point for the Macaron middleware invocation.
func (invoke MetricsInvoker) Invoke(params []interface{}) ([]reflect.Value, error) {
	invoke(params[0].(*macaron.Context), params[1].(*log.Logger))
	return nil, nil
}

// Metrics returns a middleware handler that records request metrics for exposure to Prometheus.
func Metrics() macaron.Handler {
	return func(ctx *macaron.Context, log *log.Logger) {
		start := time.Now()

		// Let the other handlers run. Items above this line happen at the beginning of a request.
		// Items below happen at the end of a request.
		ctx.Next()

		// Record the number of milliseconds the request has taken (divide by 1000000 to go from nano -> milliseconds).
		timeSince := time.Since(start).Nanoseconds() / 100000
		incrementRequestCount(ctx.Req.RequestURI, ctx.Req.Method, ctx.Resp.Status())
		recordResponseTime(ctx.Req.RequestURI, ctx.Req.Method, float64(timeSince))
	}
}

func incrementRequestCount(path, method string, responseCode int) {
	requestLabelVals := prometheus.Labels{
		"path":   path,
		"method": method,
		"status": strconv.Itoa(responseCode),
	}

	requestMetricVec, err := requests.GetMetricWith(requestLabelVals)
	if err != nil {
		log.Println("Could not record request metrics")
	}

	requestMetricVec.Inc()
}

func recordResponseTime(path, method string, responseTime float64) {
	responseLabelVals := prometheus.Labels{
		"path":   path,
		"method": method,
	}

	responseMetricVec, err := responses.GetMetricWith(responseLabelVals)
	if err != nil {
		log.Println("Could not record response metrics")
	}

	responseMetricVec.Observe(responseTime)
}
```

Import "github.com/prometheus/client_golang/prometheus/promhttp" and use the middleware like so:

```go
m.Use(handlers.Metrics())
```

And then expose the metrics like this:

```go
m.Get("/metrics", promhttp.Handler())
```

**Comment by Unknwon** [comment]
Thanks for the update!

Looks like a very good addition to the [toolbox](https://github.com/go-macaron/toolbox) package since it already does something request statistics.

**Comment by cweagans** [comment]
For what it's worth, we started using the toolbox package, but we ended up just switching entirely to this middleware in lieu of that package. The statistics that the toolbox package collects are useful, but they aren't really machine readable, which is what we needed. We were also a little wary of deploying something that would allow someone to turn on profiling in production through a URL endpoint.
